src/tinycamd.1
.TH tinycamd 1 "November 2010" "version 0.3" "USER COMMANDS"
.SH NAME
tinycamd \- an efficient HTTP interface to a webcam
.SH SYNOPSIS
.B tinycamd
[\-hv] [\-d /dev/video] [\-s 640x480] [\-p 0.0.0.0:8080]
.SH DESCRIPTION
Tinycamd is built to run efficiently, especially on very small
hardware.  Ideally you will use a camera that can provide an MJPEG
or JPEG encoded image stream. tinycamd will reformat it into JPEG images
without serious computation.  Alternativly you can use a YUYV webcam
and let tinycamd do the remainder of the encoding.
.PP
Typically, you will use a webserver to proxy to tinycamd for your
images. See the --url-prefix option for removing the remaining URL
prefix in this case.
.SH URLS
tinycamd responds to the following URLs (and some others for internal 
reasons):
.TP
/image.jpg
Return the next frame as a JPEG image. Any URL query string
will be ignored, so you can use that to defeat overzealous proxies.
.TP
/setup.html
Display a page with the camera controls exposed to HTML-5 
compatible browsers. Handy for exploring control functions.
.TP
/controls
Return an XML document defining all of the controls, their
ranges, and current values.
.TP
/set?CID=VALUE
Set a control. CID and VALUE are both integers and will have been
concocted by you with reference to the /controls URL.
.PP
... describe the UVC controls interface here...
.SH OPTIONS
.TP
\-h, \-\-help
Display a short help text.
.TP
\-d, \-\-device PATH
Specify the video device. The default is /dev/video0.
.TP
\-p, \-\-port [ADDR:]PORT
Specify the TCP port to listen on. The default is 0.0.0.0:8080, this
is port 8080 on all available interfaces. You could use
127.0.0.1:8080 if you only wanted to listen on the loopback interface to
restrict access to the local host.
.TP
\-d, \-\-daemon
Detach and run as a daemon.
.TP
\-s, \-\-size WIDTHxHEIGTH
Specify the dimensions of the video capture. The default is 640x480.
Remember that as the capture size goes up, the frame rates possible
will go down for most cameras.
.TP
\-q, \-\-quality QUALITY
Specify the JPEG encoding quality from 1 to 100. Not all cameras which
encode to JPEG will respect this, and those that do will probably have
discrete levels which they support. Likewise, MJPEG cameras generally
ignore this as well. You can use yuyv format and encode in the CPU if
you need control, but this will consume many times more CPU time.
.TP
\-f, \-\-fps NUM
The number of frames per second to capture. This will be used as a
suggestion to the camera, but it could be faster or slower depending
on what the device supports.
.TP
\-U, \-\-url-prefix PATH
If specified, this path will be removed from the front of each
URL. This is useful when you are behind a proxy that passes the
original path to tinycamd.
.TP
\-P, \-\-probe
Probe the camera and print out its capabilites. Most useful for
learning a devices supported image formats, resolutions, and frame rates.
.TP
\-F, \-\-format FMT
Set the camera's capture format. Supported values are 'jpeg', 'mjpeg',
and 'yuyv'.
.TP
\-M, \-\-monochrome
Tell the camera to capture monochrome images.
.TP
\-I, \-\-pid PATH
A path in which to write our PID. Used for daemon scripts to make
stopping easier. This must be writable by the process as invoked, it
need not be writable after the setuid and chroot.
.TP
\-i, \-\-uid NAME
After starting up and gaining access to the device and network, setuid
to this uid.
.TP
\-C, \-\-chroot PATH
After starting up, opening the camera, and recording our PID, chroot
to this path. All libraries will be loaded, it can be an empty
directory.
.TP
\-\-password ACCOUNT:PASSWORD
If specified, then HTTP Basic Authorization will be required to access
the image data. The account and password are specified and separated
by a colon. They will be hidden in 'ps' listings, but there is a
window of time immediately after launch when they will be visible.
.TP
\-\-setup-password ACCOUNT:PASSWORD
If specified, then HTTP Basic Authroization will be required to
control the camera. This account will also grant access to the image
data.
.TP
\-m, \-\-mmap
Use the mmap method to read video frames. Not generally interesting.
.TP
\-r, \-\-read
Use the read method to read video frames. Not generally interesting.
.TP
\-u, \-\-userp
Use the userp method to read video frames. Not generally interesting.
.TP
\-v, \-\-verbsose
Display a stream of debugging information.
.SH EXAMPLES
.IP
.B tinycamd
\-\-daemon 
\-\-uid nobody 
\-\-chroot /var/lib/tinycamd 
\-\-device /dev/video0
\-\-password watchman:bigeyes
.SH AUTHOR
Jim Studt, jim@studt.net


debian/changelog
tinycamd (0.3) unstable; urgency=low

  * Password protection.
  * Simplify watchdogs.
  * Add /setup.html .
  * Integrate static files.
  
 -- Jim Studt <jim@studt.net>  Thu, 11 Nov 2010 15:25:52 -0600

tinycamd (0.2) unstable; urgency=low

  * Add persistent http 1.1 sessions.

 -- Jim Studt <jim@studt.net>  Tue, 02 Nov 2010 15:28:26 -0500

tinycamd (0.1) unstable; urgency=low

  * Initial Release.

 -- Jim Studt <jim@studt.net>  Tue, 26 Oct 2010 13:30:23 -0500


tinycamd.c
/*
 * threaded.c -- A simple multi-threaded HTTPD application.
 */

#include <pthread.h>
#include <sys/types.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <sys/wait.h>
#include <stdio.h>
#include <errno.h>
#include <jpeglib.h>
#include <pwd.h>

#include "tinycamd.h"
#include "httpd.h"

#define MAXFRAME 60

extern char setup_html[];
extern int setup_html_size;
extern char tinycamd_js[];
extern int tinycamd_js_size;
extern char tinycamd_css[];
extern int tinycamd_css_size;

static void do_status_request( HTTPD_Request req)
{
  const char *status = "<html><head><title>Status</title></head>"
    "<body>status</body>"
    "</html>";

  HTTPD_Add_Header( req, "Content-type: text/html");
  HTTPD_Send_Body(req, status,strlen(status));
}

#if 0
static void put_image(const struct chunk *c, void *arg)
{
  HTTPD_Request req = (HTTPD_Request)arg;
  int i,s=0;

  for ( i = 0; c[i].data != 0; i++) s += c[i].length;

    FCGX_FPrintF(req->out, 
		 "Content-type: image/jpeg\r\n"
		 "Content-length: %d\r\n"
		 "\r\n", s);
    for ( i = 0; c[i].data != 0; i++) {
	FCGX_PutStr( c[i].data, c[i].length, req->out);
    }
    fprintf(stderr,"image size = %d\n",s);
}
#endif

static void put_single_image(const struct chunk *c, void *arg)
{
  HTTPD_Request req = (HTTPD_Request)arg;
  int i,s=0;

  HTTPD_Add_Header(req, "Cache-Control: no-cache");
  HTTPD_Add_Header(req, "Pragma: no-cache");
  HTTPD_Add_Header(req, "Expires: Thu, 01 Dec 1994 16:00:00 GMT");
  HTTPD_Add_Header(req, "Content-type: image/jpeg");

  switch(camera_method) {
    case CAMERA_METHOD_MJPEG:
    case CAMERA_METHOD_JPEG:
	{
	    unsigned char *buffer, *b;
	    for ( i = 0; c[i].data != 0; i++) s += c[i].length;
	    buffer = malloc( s);
	    for ( i = 0, b = buffer; c[i].data != 0; i++) {
		memcpy( b, c[i].data, c[i].length);
		b += c[i].length;
	    }
	    HTTPD_Send_Body( req, buffer, s);
	    free(buffer);
	}
      break;
    case CAMERA_METHOD_YUYV:
	{
	    unsigned char *jpegBuffer;
	    unsigned int jpegLeft = 1024*1024;
	    unsigned int jpegSize = 0;
	    struct jpeg_compress_struct cinfo = { .dest = 0};
	    struct jpeg_destination_mgr dmgr;
	    struct jpeg_error_mgr err;

	    jpegBuffer = malloc(jpegLeft);
	    if ( !jpegBuffer) fatal_f("Failed to allocate JPEG encoding buffer.\n");

	    void init_destination(j_compress_ptr cinfo) {
		struct jpeg_destination_mgr *d = cinfo->dest;
		d->next_output_byte = jpegBuffer;
		d->free_in_buffer = jpegLeft;
	    }
	    int empty_output_buffer(j_compress_ptr cinfo) {
		//struct jpeg_destination_mgr *d = cinfo->dest;
		log_f("eob\n");
		return  TRUE;
	    }
	    void term_destination(j_compress_ptr cinfo) {
		struct jpeg_destination_mgr *d = cinfo->dest;
		log_f("termdest\n");
		jpegSize = d->next_output_byte - jpegBuffer;
	    }
	    dmgr.init_destination = init_destination;
	    dmgr.empty_output_buffer = empty_output_buffer;
	    dmgr.term_destination = term_destination;

	    cinfo.err = jpeg_std_error(&err);
	    jpeg_create_compress(&cinfo);
	    cinfo.image_width = video_width;
	    cinfo.image_height = video_height;
	    if ( mono) {
		cinfo.input_components = 1;
		cinfo.in_color_space = JCS_GRAYSCALE;
	    } else {
		cinfo.input_components = 3;
		cinfo.in_color_space = JCS_YCbCr;
	    }
	    jpeg_set_defaults(&cinfo);
	    jpeg_set_quality(&cinfo, quality, TRUE);
	    cinfo.dest = &dmgr;

	    jpeg_start_compress( &cinfo, TRUE);
	    {
		const unsigned char *b = c[0].data;
		int row = 0;
		int col = 0;
		JSAMPLE pix[video_width*3];
		JSAMPROW rows[] = { pix};
		JSAMPARRAY scanlines = rows;

		for ( row = 0; row < video_height; row++) {
		    JSAMPLE *p = pix;
		    for ( col = 0; col < video_width; col+=2) {
			*p++ = b[0];
			if ( !mono) {
			    *p++ = b[1];
			    *p++ = b[3];
			}
			*p++ = b[2];
			if ( !mono) {
			    *p++ = b[1];
			    *p++ = b[3];
			}
			b += 4;
		    }
		    jpeg_write_scanlines( &cinfo, scanlines, 1);
		}
	    }
	    jpeg_finish_compress( &cinfo);

	    HTTPD_Send_Body( req, jpegBuffer, jpegSize);
	    free(jpegBuffer);
	}
      break;
  }

  log_f("image size = %d\n",s);
}

#if 0
static void stream_image( HTTPD_Request req)
{
    int i;
    const char *boundary = "--myboundary\r\n";

    HTTPD_Add_Header( req, "Cache-Control: no-cache");
    HTTPD_Add_Header( req, "Pragma: no-cache");
    HTTPD_Add_Header( req, "Expires: Thu, 01 Dec 1994 16:00:00 GMT");
    HTTPD_Add_Header( req, "Content-Type: text/xml");
    HTTPD_Add_Header( req, "Content-Type: multipart/x-mixed-replace; boundary=--myboundary");

    for(i = 0; i < MAXFRAME; i++) {
      HTTPD_Send_Body_Chunk( req, boundary, sizeof(boundary)-1);
      with_next_frame( &put_image, req);
      HTTPD_Push(req);
    }
}
#endif

static void do_video_call( HTTPD_Request req, video_action action, int cid, int val)
{
    char buf[8192];

    with_device( action, buf, sizeof(buf), cid, val);
    HTTPD_Add_Header( req, "Cache-Control: no-cache");
    HTTPD_Add_Header( req, "Pragma: no-cache");
    HTTPD_Add_Header( req, "Expires: Thu, 01 Dec 1994 16:00:00 GMT");
    HTTPD_Add_Header( req, "Content-Type: text/xml");
    HTTPD_Send_Body( req, buf, strlen(buf));
}

static int demand_authorization(HTTPD_Request req)
{
    HTTPD_Send_Status( req, 401, "Authorization Required");
    HTTPD_Add_Header( req, "WWW-Authenticate: Basic realm=\"tinycamd\"");
    HTTPD_Send_Body(req, "login required", 14);
    return 0;
}

static int check_password( HTTPD_Request req, int setup) {

    if ( setup ) {
	if ( !setup_password) return 1;
    } else {
	if ( !setup_password && !password) return 1;
    }

    {
	const char *auth = HTTPD_Get_Authorization(req);

	if ( !auth) return demand_authorization(req);

	// a setup_password will do anything, if we have it.
	if ( setup_password && strcmp( auth, setup_password) == 0) return 1;
	if ( setup) return demand_authorization(req);

	if ( password && strcmp( auth, password)==0) return 1;
    }
    return demand_authorization(req);
}

static void handle_requests(HTTPD_Request req, const char *method, const char *rawUrl)
{
  int cid,val;
  const char *url = rawUrl;

  if ( strncmp( rawUrl, url_prefix, strlen(url_prefix))) {
      url = "***BADURL-NOPREFIX***";
  } else {
      url = rawUrl + strlen(url_prefix);
  }

  log_f("Request: %s %s => %s\n", method, rawUrl, url);
  if ( strcmp(url,"/status")==0) {
    do_status_request(req);
  } else if ( strcmp(url,"/setup.html")==0) {
      if ( check_password(req, 1)) {
	  HTTPD_Add_Header( req, "Content-type: text/html");
	  HTTPD_Send_Body(req, setup_html,setup_html_size);
      }
  } else if ( strcmp(url,"/tinycamd.js")==0) {
      HTTPD_Add_Header( req, "Content-type: text/javascript; charset=utf8");
      HTTPD_Send_Body(req, tinycamd_js,tinycamd_js_size);
  } else if ( strcmp(url,"/tinycamd.css")==0) {
      HTTPD_Add_Header( req, "Content-type: text/css");
      HTTPD_Send_Body(req, tinycamd_css,tinycamd_css_size);
#if 0
  } else if ( strcmp(url,"/image.replace")==0) {
    stream_image(req);
#endif
  } else if ( strcmp(url,"/controls")==0) {
    do_video_call( req, list_controls,0,0);
  } else if ( sscanf(url,"/set?%d=%d",&cid,&val)==2 ) {
    if ( check_password(req,1)) do_video_call( req, set_control,cid,val);
  } else if ( strcmp(url,"/")==0 ||
	      strcmp( url, "/image.jpg") == 0 ||
	      strncmp( url, "/image.jpg?", 11) == 0) {
      if ( check_password(req, 0)) with_current_frame( &put_single_image, req);
  } else {
    HTTPD_Send_Status( req, 404, "Not Found");
    HTTPD_Send_Body( req, "404 - Not found", 15);
  }
}

static void *sleeper(void *arg)
{
  while(1) sleep(1000);
  return 0;
}

int main(int argc, char **argv)
{
    pthread_t captureThread;
    pthread_t httpdThread;

    do_options(argc, argv);

    if ( daemon_mode) {
      if ( daemon(0,0) == -1) {
	fatal_f("Failed to become a daemon: %s\n", strerror(errno));
      }
    }

    if ( pid_file) {
	FILE *pf = fopen( pid_file,"w");
	if ( !pf) fatal_f("Failed to open pid file %s: %s\n", pid_file, strerror(errno));
	fprintf(pf, "%d\n", getpid());
	if ( fclose(pf)==EOF) fatal_f("Failed to close pid file %s: %s\n", pid_file, strerror(errno));
    }

    open_device();

    if ( probe_only) {
	probe_device();
	return 0;
    }

    init_device();
    start_capturing();

    pthread_create( &captureThread, NULL, main_loop, NULL);

    /*
    ** I am so sorry. But glibc dynamically loads libgcc_s.so.1 to handle pthread_cancel, so
    ** I need to get that in before chrooting.
    */
    {
      pthread_t crappyHack;
      void *whatever;

      if ( pthread_create( &crappyHack, 0, sleeper, 0)) fatal_f("Failed to start test thread.\n");
      if ( pthread_cancel( crappyHack)) fatal_f("Failed to cancel test thread.\n");
      if ( pthread_join( crappyHack, &whatever)) fatal_f("Failed to join test thread.\n");
    }

    /*
    ** Slink into our ghetto and lower our privileges in preparation for handling queries.
    */
    {
        uid_t uid = getuid();
	
        /* lookup uid before we chroot... or it is gone. */
        if ( setuid_to) {
            struct passwd *pw = getpwnam(setuid_to);
            if ( !pw) fatal_f("Failed to lookup user `%s' for setuid: %s\n", setuid_to, strerror(errno));
            uid = pw->pw_uid;
        }

        if ( chroot_to) {
            if ( chroot( chroot_to)) fatal_f("Failed to chroot to `%s': %s\n", chroot_to, strerror(errno));
        }

        if ( setuid_to) {
            if ( setreuid(uid, uid)) fatal_f("Failed to setuid to `%s': %s\n", setuid_to, strerror(errno));
        }
    }

    httpdThread = HTTPD_Start( bind_name, handle_requests);

    for(;;) sleep(100);

    close_device();

    return 0;
}
